concurrency/executor.go

```diff
@@ -0,0 +1,212 @@
+
+// ChannelWithResult is an interface that should be implemented by the result channel
+type ChannelWithResult[ResultType any] interface {
+	GetResult() ResultType
+}
+
+// GetResults returns all errors that occurred during processing
+func (e *ConcurrentExecutor[ResultType, ResultChannelType, TaskType]) GetErrors() []error {
+	return e.errors
+}
+
+// ExecuteSimple executes a task that doesn't require a payload. It is executed repeatTimes times with given concurrency. The simpleProcessorFn is the function that processes the task.
+// Executor will attempt to distribute the tasks evenly among the executors.
+func (e *ConcurrentExecutor[ResultType, ResultChannelType, TaskType]) ExecuteSimple(concurrency int, repeatTimes int, simpleProcessorFn SimpleTaskProcessorFn[ResultChannelType]) ([]ResultType, error) {
+	dummy := make([]TaskType, repeatTimes)
+	for i := 0; i < repeatTimes; i++ {
+		dummy[i] = *new(TaskType)
+	}
+
+	return e.Execute(concurrency, dummy, adaptSimpleToTaskProcessorFn[ResultChannelType, TaskType](simpleProcessorFn))
+}
+
+// Execute executes a task that requires a payload. It is executed with given concurrency. The processorFn is the function that processes the task.
+// Executor will attempt to distribute the tasks evenly among the executors.
+func (e *ConcurrentExecutor[ResultType, ResultChannelType, TaskType]) Execute(concurrency int, payload []TaskType, processorFn TaskProcessorFn[ResultChannelType, TaskType]) ([]ResultType, error) {
+	if len(payload) == 0 {
+		return []ResultType{}, nil
+	}
+
+	if concurrency <= 0 {
+		e.logger.Warn().Msg("Concurrency is less than 1, setting it to 1")
+		concurrency = 1
+	}
+
+	var wgProcesses sync.WaitGroup
+	wgProcesses.Add(len(payload))
+
+	canSafelyContinueCh := make(chan struct{}) // waits until listening goroutine finishes, so we can safely return from the function
+	doneProcessingCh := make(chan struct{})    // signals that both result and error channels are closed
+	errorCh := make(chan error, len(payload))
+	resultCh := make(chan ResultChannelType, len(payload))
+
+	// mutex to protect shared state
+	mutex := sync.Mutex{}
+
+	// atomic counter to keep track of processed tasks
+	var atomicCounter atomic.Int32
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// listen in the background until all tasks are processed (no fail-fast)
+	go func() {
+		defer func() {
+			e.logger.Trace().Msg("Finished listening to task processing results")
+			close(canSafelyContinueCh)
+		}()
+		for {
+			select {
+			case err, ok := <-errorCh:
+				if !ok {
+					e.logger.Trace().Msg("Error channel closed")
+					return
+				}
+				if err != nil {
+					mutex.Lock()
+					e.errors = append(e.errors, err)
+					e.logger.Err(err).Msg("Error processing a task")
+					mutex.Unlock()
+					wgProcesses.Done()
+
+					// cancel the context if failFast is enabled and it hasn't been cancelled yet
+					if e.failFast && ctx.Err() == nil {
```

Comment:
what does failing mean in this context? we already add error the executor and then cancel the context. Cancelling it marks all tasks that hasn't been submitted for processing as done. Here I am not existing the goroutine, because there might be some tasks that are in flight (already submitted), and if I exited this goroutine they would never call `wgProcess.Done()` and `wgProcesses.Wait()` would block forever.

Is that what you meant by failing or something else?

Reply:
I was thinking from a perspective where context is passed from a parent, in that case if the parent context fails, should it fail as well?

Reply:
And yes I meant exiting goroutine, I see your logic